### ISSUE TITLE
Effect of throw expressions on reachability is unspecified

### ISSUE THREAD

**Issue by sharwell** [dotnet/csharplang#656]
As described in the commentary on dotnet/roslyn#19984, the impact of throw expressions on reachability analysis is not currently clear. Currently the specification for [throw expressions](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.0/throw-expression.md) only says this:

> Note: the rest of the semantics of the throw expression are identical to the semantics of the *throw_statement* in the current language specification.

I believe that the impact of this new construct on reachability should be explicit rather than implicit, and propose the following changes as part of the Throw expression feature.

### Syntax

The following is a more complete representation of the intended syntax of this feature.

```diff
+throw_expression
+    : 'throw' expression
+    ;
 
 null_coalescing_expression
     : conditional_or_expression
     | conditional_or_expression '??' null_coalescing_expression
+    | conditional_or_expression '??' throw_expression
     ;
 
 conditional_expression
     : null_coalescing_expression
-    | null_coalescing_expression '?' expression ':' expression
+    | null_coalescing_expression '?' expression_or_throw ':' expression_or_throw
     ;
 
+expression_or_throw
+    : expression
+    | throw_expression
+    ;
 
 anonymous_function_body
-    : expression
+    : expression_or_throw
     | block
     ;

 arrow_expression_body // NOTE this doesn't actually appear to exist anywhere yet
-    : '=>' expression ';'
+    : '=>' expression_or_throw ';'
     ;
```

#### Intent

The intent of the semantics for a throw_expression include the expansion of the concept of end point reachability to expressions. The end point of an expression is, in general, reachable unless evaluation of the complete expression necessarily includes the evaluation of a *throw_expression*. For the purposes of this determination, the value of constant expressions are considered for the purposes of determining reachability for conditionally-executed expressions, including:

* The ternary operator, where the condition may be a constant
* Null coalescing expressions, where the left hand side may be a constant
* Conditional logical operators (short-circuiting), where the left hand side may be a constant
* Null conditional operators (`?.` and `?[`)

:memo: While expressions have start and end point reachability as a result of this change, a warning for unreachable subexpressions may be omitted for compatibility purposes. For example, the following code was legal prior to C# 7, and did not report a warning related to reachability:

```csharp
var ignored = true ? Reachable() : Unreachable();
```

#### Details

⚠️ The following is really just notes for now.

* *declaration_statement*
  * The end point of a *declaration_statement* is reachable if the end point of its last *local_variable_initializer* is reachable. If the *declaration_statement* does not include any initializers, then the end point is reachable if the start point is reachable.
* *expression_statement*
  * TODO
* *local_variable_initializer*
  * TODO
* *expression*
  * Unless otherwise specified, the end point of an *expression* is reachable if the start point is reachable.
* *array_initializer*
  * TODO
* *local_variable_initializer_unsafe*
  * TODO
* *throw_expression*
  * The end point of a *throw_expression* is never reachable
* *invocation_expression*
  * The end point of an *invocation_expression* is reachable if it has no *argument_list*, or if the end point of its *argument_list* is reachable.
* *element_access*
  * The end point of an *element_access* is reachable if the end point of its *expression_list* is reachable
*

**Comment by CyrusNajmabadi** [comment]
>While expressions have start and end point reachability as a result of this change, a warning for unreachable subexpressions may be omitted for compatibility purposes. 

A good option for a warning wave.

Note: In at least one versoin of C# we did have expression reachability warnings.  But we removed it (maybe because it was not per spec).

**Comment by gafter** [comment]
The intent of the specification was not to define any new behavior for the reachability of expressions. The intent of that reference to the throw statement was to explain the handling and type constraints of the thrown expression.

If we have a concept of expression reachability, but it has no effect on the behavior of the compiler, why would we bother?

Moreover, why would we do it for a *throw-expression*, but not similarly warn for this kind of code?

``` c#
int x;
if (false && M())
{
    Console.WriteLine(x); // formally considered reachable in C# 1.0, but x is definitely assigned
}
```

**Comment by sharwell** [comment]
> Moreover, why would we do it for a *throw-expression*, but not similarly warn for this kind of code?

For compatibility.

**Comment by gafter** [comment]
> For compatibility

Since throw expressions are already in the language, we shouldn't warn in situations involving them either.

**Comment by sharwell** [comment]
@gafter Since the compat bar tends to differ a bit between things released in C# 1 and those in C# 7, I felt the proposal would be more reasonably accepted if it only impacted a scenario first introduced in the latter. I would see the expansion of expression reachability concept to other cases as a follow-up issue to this, and one likely blocked on some warning waves functionality.

**Comment by CyrusNajmabadi** [comment]
I rather us just embrace the concept of expression-reachability fully and add warnings for unreachable expressions.  Because of the compat concern, this would have to come with something like warning waves/strict-mode to enable them.  

If we do that, then throw-expressions nicely slot in, and we're providing more general value for existing code that has shipped and which would benefit from warning for code we can tell is wonky.

**Comment by gafter** [comment]
There is a way of doing a comprehensive extension of reachability to expressions, but the approach you suggest doesn't work very well (it gets very complex). On the other hand, we could remove all of the reachability rules from the language, and replace them by this simple specification that also handles the reachability of expressions:

> To determine whether a particular statement or end point is reachable, the compiler performs flow analysis according to the definite assignment rules in section 5.3. If a hypothetical variable `unreachable` that is not definitely assigned at the beginning of the method body is found to be definitely assigned at the beginning of a statement, then that statement is unreachable.  Conversely, if `unreachable` is not definitely assigned at the beginning of a statement, then the statement is reachable.  Similarly, the end point of a statement is reachable if and only if `unreachable` is not definitely assigned after that statement.

Plus we need to remove the rule that says that all variables are definitely assigned in unreachable code to keep this rule from being circularly defined.

Unfortunately, it would cause warnings not produced today. The principal incompatibility caused by this approach is that this takes into account the values of constant expressions (§7.19) that control the behavior of statements, while the current spec for reachability does not.

However, I don't know *why* we would do any of this. Who benefits from it?

**Comment by sharwell** [comment]
@gafter I would be in favor of language like that primarily due to the resulting simplicity in the language specification. If we change the language regarding unreachable statements as follows, then it could be considered an achievable goal:

> A warning ~~is~~ **may be** reported if the compiler determines that a statement **or expression** is unreachable. It is specifically not an error for a statement **or expression** to be unreachable.

**Comment by alrz** [comment]
>I don't know why we would do any of this. Who benefits from it?

Somehow related: some code patterns were previously mentioned that could take advantage of a more advanced definitive assignment rules:
```cs
if (obj is Foo foo is var result) {
  // foo is not definitively assigned but could be
}

// equivalent to
bool result;
if (result = Foo foo is result)

// or with declaration expressions
if (var result = obj is Foo foo)

// likewise
if (map?.TryGetValue(key, out var result) != null) { 
  // result is not definitively assigned but could be
}
```

**Comment by sharwell** [comment]
> However, I don't know *why* we would do any of this. Who benefits from it?

Reachability and definite assignment are two different but related methods of moving particular kinds of program error detection from runtime to compile time. On the assumption that the former is valuable, this proposal represents a natural extension of a long-existing C# concept to a new C# feature. In addition, in my opinion this extension is reasonably inferred from the incomplete specification provided for the new feature to date, and thus seeks to explicitly state/clarify behavior which I already expected to see.

**Comment by CyrusNajmabadi** [comment]
I agree with Sam on this.  As a user, i would benefit from having the compiler tell me early that something isn't quite right with my program.  I already like and reachability warnings for statements useful, and all the reasons i like it there apply to expressions as well.

**Comment by gafter** [comment]
> ``` c#
if (obj is Foo foo is var result) {
  // foo is not definitively assigned but could be
}
```

No, it could not be. This is equivalent to

``` c#
var tmp = obj is Foo foo;
if (tmp is var result) {
  // foo is not definitely assigned when tmp is false
}
```

**Comment by alrz** [comment]
Yes, should've added a `&& result` to the condition which still won't work probably because we don't look inside descendant nodes to determine definitive assignments.

**Comment by gafter** [comment]
The "definitely assigned when true" state is not carried with a boolean variable.

**Comment by gafter** [comment]
This has now been fixed. The specification has been clarified.

If you're still interested in getting warnings for unreachable expressions, please open a new issue for that.
